crates/holochain/src/core/workflow/integrate_dht_ops_workflow.rs

```diff
@@ -17,100 +45,126 @@ pub async fn integrate_dht_ops_workflow(
+            )?;
+
+            set_ops_to_integrated(
+                txn,
+                SET_ADD_LINK_OPS_TO_INTEGRATED,
+                named_params! {
                     ":when_integrated": time,
                     ":create_link": ChainOpType::RegisterAddLink,
-                })?;
-            total += changed;
-            let changed = txn
-                .prepare_cached(holochain_sqlite::sql::sql_cell::UPDATE_INTEGRATE_DEP_STORE_RECORD)?
-                .execute(named_params! {
+                },
+                &mut stored_ops,
+            )?;
+
+            set_ops_to_integrated(
+                txn,
+                UPDATE_INTEGRATE_DEP_STORE_RECORD,
+                named_params! {
                     ":when_integrated": time,
                     ":store_record": ChainOpType::StoreRecord,
                     ":updated_record": ChainOpType::RegisterUpdatedRecord,
                     ":deleted_by": ChainOpType::RegisterDeletedBy,
-                })?;
-            total += changed;
-            let changed = txn
-                .prepare_cached(holochain_sqlite::sql::sql_cell::SET_DELETE_LINK_OPS_TO_INTEGRATED)?
-                .execute(named_params! {
+                },
+                &mut stored_ops,
+            )?;
+
+            set_ops_to_integrated(
+                txn,
+                SET_DELETE_LINK_OPS_TO_INTEGRATED,
+                named_params! {
                     ":when_integrated": time,
                     ":create_link": ChainOpType::RegisterAddLink,
                     ":delete_link": ChainOpType::RegisterRemoveLink,
 
-                })?;
-            total += changed;
-            let changed = txn
-                .prepare_cached(
-                    holochain_sqlite::sql::sql_cell::SET_CHAIN_INTEGRITY_WARRANT_OPS_TO_INTEGRATED,
-                )?
-                .execute(named_params! {
+                },
+                &mut stored_ops,
+            )?;
+
+            set_ops_to_integrated(
+                txn,
+                SET_CHAIN_INTEGRITY_WARRANT_OPS_TO_INTEGRATED,
+                named_params! {
                     ":when_integrated": time,
                     ":chain_integrity_warrant": WarrantOpType::ChainIntegrityWarrant,
-                })?;
-            total += changed;
-            WorkflowResult::Ok((total, activity_to_integrate))
+                },
+                &mut stored_ops,
+            )?;
+
+            WorkflowResult::Ok((activity_to_integrate, stored_ops))
         })
         .await?;
     // Once the database transaction is committed, update the cache with the
     // integrated activity.
     dht_query_cache
         .set_all_activity_to_integrated(activity_integrated)
         .await?;
+    let changed = stored_ops.len();
     let ops_ps = changed as f64 / start.elapsed().as_micros() as f64 * 1_000_000.0;
     tracing::debug!(?changed, %ops_ps);
     if changed > 0 {
+        network.new_integrated_data(stored_ops).await?;
```

Comment:
Whoo!

Reply:
This logic that `if something was integrated, it's incomplete, so run again, and otherwise if nothing was integrated, it's complete` has to change, but I'll leave that for the refactor. I tried to strike the balance between improving obvious things here and just sticking to making the required change.

Reply:
Good choice. This K2 integration is already massive and finding bugs in it could be tough. Let's stabilize that and then come back for more :)